### DISCUSSION TITLE
Hitting a bit of a wall

### DISCUSSION THREAD

**Discussion by osotsia** [KellerJordan/modded-nanogpt#85] [General]
This speedrun is looking like a very good 10x cumulative improvement so far. But it seems to be hitting a bit of a wall.

![Figure_1](https://github.com/user-attachments/assets/f01ea0a7-db28-4461-9fa6-c312461cdf17)

I'm trying to think ahead. Is it correct to say that since AI performance depends on the log of compute, we'd have to hit a 100x cumulative improvement for the next step up? So 31 min -> 3 min -> 0.3 min? Because if so, that might require something truly radical.

**Comment by linux-leo** [comment]
I agree, knowing that the effort required to get a substantial record is now so high, I'm struggling to find motivation.

I know suggestions are meaningless without someone doing the work, and every "radical" Idea I've had looks like so much work that I'm scared of blowing through my budget multiple times over. But let me restate them, because I think they are worth pursuing:

1. Forgetting Transformer: Softmax Attention with a Forget Gate: https://openreview.net/forum?id=q2Lnyegkr8 (See: https://github.com/KellerJordan/modded-nanogpt/discussions/23#discussioncomment-11878437)

2. Exploring the Benefit of Activation Sparsity in Pre-training: https://arxiv.org/abs/2410.03440


I think the best option would be for someone to implement kv shifting as outlined in the forgetting transformer. It's (hopefully) a performance improvement by itself and with it, all pre-requisites for implementing the forgetting transformer are fullfilled. And it looks relatively simple, there are other papers that have suggested kv shifting and even provided reference code: https://arxiv.org/pdf/2411.19574

**Reply by osotsia** [reply]
I just did some quick KV shifting work. It was better at step 125 of training: val loss = 4.55 vs 4.65 in the current leader (RuleTweak). But RuleTweak quickly made up ground, and they ended at the same place. My observation is the improvements in that Baichuan AI paper are not that big (# steps saved at the same horizontal loss value Fig 4,6). Good instincts though, it's still the best experiment I've done so far. Will work on the rest and let you know if I find something.

**Reply by linux-leo** [reply]
Nice Work @osotsia! Two suggestions:

1. Applying qk norm after shifting like it was done in the forgetting transformer might improve performance. Idk If RoPE should be applied before or after shifting, but might as well also put it after shifting so positional information isn't lost.

2. The Baichuan paper implements data independent kv shifting, while RWKV and the forgetting transformer use data dependent kv shifting, which should be even better, see Equation 14 and Appendix A of the Forgetting Transformer paper.

**Reply by osotsia** [reply]
Did those suggestions. Same pattern in the val losses.

I think I'll go straight for the forget gate. The ablations in the forgetting transformer paper (Table 3) indicates it might have the biggest gains. Those tend to be more transferable to different contexts. I'll pick it up next week.

**Reply by linux-leo** [reply]
Yeah I was following the commits in your fork  @osotsia. Great work. From the multiple versions you committed, which ones did you test and which had the lowest loss in the earlier training steps? I just want to know the best trade off between the complexity of the implementation (and it's overhead) and the validation loss.

I know forgetting transformers are the biggest performance improvement, but combining it with kv shifting methods could allow removing RoPE, which would further reduce runtime, similar to how it was done in the forgetting transformer paper, just without output norm and gate.

**Reply by osotsia** [reply]
Data-independent [09eba8b](https://github.com/osotsia/modded-nanogpt-refactor/commit/09eba8b9a0486314bb3b655986d0260f35ffbb1a)

Data-dependent shift [680e1de](https://github.com/osotsia/modded-nanogpt-refactor/commit/680e1deb339118ac5de965d4ce4653789871674b)

Those are the working tested ones (the other commits usually had something wrong with them). They performed basically the same (in terms of early val loss reduction), but yeah the data-dependent one was slower. You can just grab the class CausalSelfAttention if you want.

**Reply by osotsia** [reply]
I'm testing out convolutions (as something more general than KV shift) but I'm running into flexattention bugs (I think). Before I dive too deep into this, I just wanted to check that this has been tried? I'm testing out MDHA from the Primer architecture paper from google, same one with relu squared.

**Reply by linux-leo** [reply]
That's a great idea @osotsia and something I thought of myself, but I didn't try. As for flex attention bugs, I unfortunately don't know anything about the underlying cuda/triton kernels, maybe you can open an issue in the pytorch repo or the [attention-gym repo](https://github.com/pytorch-labs/attention-gym)

But on a side note, the forgetting transformer paper has been released recently and with it the code for the kv shifting and forget mechanism: https://github.com/zhixuan-lin/forgetting-transformer/blob/main/src/forgetting_transformer/model/forgetting_transformer/modeling_forgetting_transformer.py

And the kv shifting has a few changes compared to your implementation. For example I think for the kv shifting they keep separate parameter for each attention head compared to a single parameter per attention block in your implementation, so with the changes from the official codebase you might be able to squeeze more performance out of the kv shifting mechanism and use it as an alternative to convolutions.

**Reply by koute** [reply]
> As for flex attention bugs, I unfortunately don't know anything about the underlying cuda/triton kernels, maybe you can open an issue in the pytorch repo

FWIW, last time I used FlexAttention it was, well, very buggy (and I was only doing basic stuff with it!), so I wouldn't be surprised if there are still a lot of issues with it, especially if you do something more exotic. I would *definitely* not trust it without writing a unit test that compares its behavior to an equivalent eager implementation. In the end I switched back to flash attention because it works and was faster than FlexAttention.

**Comment by osotsia** [comment]
Update a year later

<img width="1524" height="1119" alt="nanogpt_trend_update" src="https://github.com/user-attachments/assets/ef821aca-eaee-41cf-9c51-77c2a3ce369f" />



So when I looked at that, I got a little worried about overfitting. This may nor may not be important. I think it's more a scientific question, and this speedrun is more like an all-out engineering challenge.

Since the speedrun completes in under 400M tokens, it only consumes training shards 1 through 4. So any later training shard (e.g., 10 or 11) can also be a validation set. Evaluating the latest code on these unseen shards reveals a persistent loss well above the 3.28 target.

**Results on 8xH100:**

```text
# Unmodified latest version (normal validation)
step:1390/1390 val_loss:3.2782 train_time:79690ms step_avg:57.33ms

# Latest code validated on fineweb_train_000010
step:1390/1390 val_loss:3.2943 train_time:79364ms step_avg:57.10ms
step:1390/1390 val_loss:3.2969 train_time:79317ms step_avg:57.06ms
step:1390/1390 val_loss:3.2956 train_time:79425ms step_avg:57.14ms

# Latest code validated on fineweb_train_000011
step:1390/1390 val_loss:3.2979 train_time:80070ms step_avg:57.60ms
step:1390/1390 val_loss:3.2961 train_time:80191ms step_avg:57.69ms
step:1390/1390 val_loss:3.2956 train_time:80267ms step_avg:57.75ms
```

I really wanted to trace the historical timeline of this gap, but I kept running into environment compatibility issues. Maybe someone with more repo familiarity (and deeper pockets) is curious?



**Reproduction Script:**

```bash
# 1. Clone the repository and enter the directory
git clone https://github.com/KellerJordan/modded-nanogpt.git
cd modded-nanogpt

# (Optional) check out a specific commit
# COMMIT="2d21801347ed173702883f59e39d96b308e6a07a"
# git stash -q
# git checkout $COMMIT -q

# 2. Install required dependencies
pip install -r requirements.txt

# 3. Download the first 11 training shards (and the original val shard)
python data/cached_fineweb10B.py 11

# 4. Use sed to replace the validation file target in train_gpt.py inline
sed -i 's|"data/fineweb10B/fineweb_val_\*.bin"|"data/fineweb10B/fineweb_train_000011.bin"|g' train_gpt.py

# 5. Execute the training script multiple times
NUM_RUNS=3
for i in $(seq 1 $NUM_RUNS); do
    echo "========================================"
    echo "Starting run $i of $NUM_RUNS..."
    echo "========================================"
    ./run.sh
done
```

**Comment by ClassicLarry** [comment]
Its possible there are 5-10 seconds or so of overfitting. I would expect at least a couple seconds. Or maybe something about the first validation doc is just easier.

**Reply by osotsia** [reply]
Well we won't know until someone gets old versions of this repo working & traces how long that gap has been around.
